### ISSUE TITLE
Error Formatting Coverage Report when xsl:variable is the last element - xSpec 3.1.2

### ISSUE THREAD

**Issue by birdya22** [xspec/xspec#2019]
Hi,
  a got a failure formatting a coverage report when I'd commented out some code and left a `xsl:variable` as the last element.
This is not a global `xsl:varible`.

Saxon produced a warning:
```
Warning at xsl:variable on line 9 column 33 of VariableLastElement.xsl:
  SXWN9001  A variable with no following sibling instructions has no effect
```

xSpec produced this:
```
Formatting Coverage Report...
Error at xsl:apply-templates on line 223 column 38 of coverage-compute-status.xsl:
  XTTE0505  An empty sequence is not allowed as the result of template match="XSLT:variable"
  mode="coverage". Found while atomizing the value of variable $coverage in
  {xsl:apply-templates} on line 256
During lazy evaluation of if($matches) then ... else ... on line 254 of file:/K:/xSpec/3_1_2/src/reporter/coverage-report.xsl
In template output-lines on line 196 column 51 of coverage-report.xsl:
     invoked by xsl:iterate at file:/K:/xSpec/3_1_2/src/reporter/coverage-report.xsl#231
In template output-lines on line 196 column 51 of coverage-report.xsl:
     invoked by xsl:call-template at file:/K:/xSpec/3_1_2/src/reporter/coverage-report.xsl#123
  In template rule with match="element(Q{http://www.w3.org/1999/XSL/Transform}stylesheet)" on line 97 of coverage-report.xsl
     invoked by xsl:apply-templates at file:/K:/xSpec/3_1_2/src/reporter/coverage-report.xsl#92
  In template rule with match="document-node(element(Q{}trace))" on line 73 of coverage-report.xsl
     invoked by xsl:apply-templates (tail calls omitted) at file:/K:/xSpec/3_1_2/src/reporter/coverage-report.xsl#64
An empty sequence is not allowed as the result of template match="XSLT:variable" mode="coverage". Found while atomizing the value of variable $coverage in {xsl:apply-templates} on line 256

*** Error formatting the coverage report

```

The issue is on line 222 of coverage-compute-status.xsl where there is an assumption that there will always be a following element.

I've attached a test case (adding.txt to the file names).
[VariableLastElement.xspec.txt](https://github.com/user-attachments/files/17677673/VariableLastElement.xspec.txt)
[VariableLastElement.xsl.txt](https://github.com/user-attachments/files/17677674/VariableLastElement.xsl.txt)

Adrian

**Comment by galtm** [comment]
Adrian, thanks for reporting this bug clearly.

To help determine what the coverage status should be in your test case, I looked at the similar case where the `xsl:variable` and `xsl:text` elements are swapped. In that case,

- The coverage status for `xsl:variable` is 'hit' due to the following sibling, even though Saxon might be optimized to skip computing the extraneous variable's value.
- Under the hood, the Saxon trace omits a hit on `xsl:variable` by default, but includes it if I set the `SAXON_CUSTOM_OPTIONS` environment variable to `-opt:0` before generating the coverage report.

I'm thinking that the fix for this bug should show a local `xsl:variable` as 'hit' when there are no following siblings other than `xsl:variable`. Do you think this makes sense?

**Comment by birdya22** [comment]
My thoughts are based on the 3rd paragraph in the [XSLT 3.0 Spec](https://www.w3.org/TR/xslt-30/#scope-of-variables) which says:

> A local variable binding element is visible for all following siblings and their descendants ...

If there are no following siblings the local variable cannot be used, because nothing can see the value, which is what the Saxon waring is saying.
Even if there are a sequence of local variables, where the first one supplies a value to the next one etc. nothing can use  the values (I don't know if Saxon would detect that and not execute any of them).

I don't think that 'hit' is appropriate, but I'm between 'missed' and 'ignored'.
The variable could be considered as a Declaration (because it isn't used) and marked as 'ignored'.
Or it could be considered as an Instruction that isn't executed and marked as 'missed'.

Based on the Saxon warning message:

> A variable with no following sibling instructions has no effect

I think I would tend to go for 'ignored'.

Adrian

**Comment by galtm** [comment]
You're right, 'hit' is not appropriate. I realized that a template could consist of `<xsl:sequence select="error(...)"/>` followed by `<xsl:variable>` and nothing else. Calling that local variable 'hit' would be just plain wrong.

I like the idea of ignoring a local variable whose scope is empty, but the way we've defined the 'ignored' coverage status in https://github.com/xspec/xspec/blob/master/docs/xslt-code-coverage-by-element.md doesn't seem to fit:

>Ignored - the node is a declaration and no trace information is provided. Without extensive work in XSpec to parse the stylesheet it is not possible to determine if these declarations are used

I don't think we should consider a local variable as a Declaration, because the XSLT 3.0 spec (start [here](https://www.w3.org/TR/xslt-30/#dt-declaration) and look under the list) says, "Note that the xsl:variable and xsl:param elements can act either as [declarations](https://www.w3.org/TR/xslt-30/#dt-declaration) or as [instructions](https://www.w3.org/TR/xslt-30/#dt-instruction). A global variable or parameter is defined using a declaration; a local variable or parameter using an instruction."

If we go with 'ignored' I think we should update the description of that status in [our document](https://github.com/xspec/xspec/blob/master/docs/xslt-code-coverage-by-element.md), rather than consider a local variable as a Declaration. The challenge is how to maintain a clear distinction between the descriptions of 'ignored' status and 'unknown' status. (Does that beg the question of whether empty-scope local variables should have 'unknown' status?)

Our description of 'missed' is, "the node was not executed" and that's consistent with Saxon not reporting a hit on that node.

I think I need to give the status some more thought.

**Comment by galtm** [comment]
For what it's worth, this bug was present in XSpec v2.3.2. If a local variable has no following siblings other than other local variables, then the line
https://github.com/xspec/xspec/blob/e650d664b4dcade4cec6f763f1cf3bf84be51908/src/reporter/coverage-report.xsl#L363

returns an error like this:
```
Error evaluating (xsl:iterate) in xsl:sequence/@select on line 363 column 115 of coverage-report.xsl:
  XPTY0004: An empty sequence is not allowed as the first argument of local:coverage().
  Found while atomizing the value of variable $coverage
at function local:coverage on line 404 of coverage-report.xsl:
     invoked by function call at file:/.../xspec/bin/../src/reporter/coverage-report.xsl#259
at template output-lines on line 200 of coverage-report.xsl:
     invoked by xsl:iterate at file:/.../xspec/bin/../src/reporter/coverage-report.xsl#236
at template output-lines on line 200 of coverage-report.xsl:
     invoked by xsl:call-template at file:/.../xspec/bin/../src/reporter/coverage-report.xsl#126
  In template rule with match="element(Q{http://www.w3.org/1999/XSL/Transform}stylesheet)" on line 95 of coverage-report.xsl
     invoked by xsl:apply-templates at file:/.../xspec/bin/../src/reporter/coverage-report.xsl#90
  In template rule with match="document-node(element(Q{}trace))" on line 71 of coverage-report.xsl
     invoked by xsl:apply-templates at file:/.../xspec/bin/../src/reporter/coverage-report.xsl#62
  In template rule with match="document-node(element(Q{}trace))" on line 61 of coverage-report.xsl
An empty sequence is not allowed as the first argument of local:coverage(). Found while atomizing the value of variable $coverage
```

That line is part of an `xsl:choose` construct, and if I conditionalize that branch to avoid the error, the branch that finally gets reached results in a coverage status of 'missed'.

I suspect that the bug has been present even longer; the code looked only slightly different in https://github.com/xspec/xspec/blob/bc166228cd46987f3a731eb51d8e9b72d10d38bd/src/reporter/coverage-report.xsl from 4 years ago (v1.6, if I'm interpreting the GitHub signals correctly from #794). I haven't actually run that code, though, or looked further back in the code history.

**Comment by birdya22** [comment]
I have to admit I hadn't read the rule about when a variable is a Declaration vs. Instruction. My reasoning was based more on my thought about the logical state of it. 

Having given it a bit more thought I am now leaning towards 'missed'. My thinking is that in general 'missed' can be turned into 'hit' by the user writing more tests, or in this case, by actually using the variable.
Based on our current definition for 'ignored', a node will always be marked as 'ignored' and the user can't change that. As in the previous sentence that isn't correct in this case, so I'm thinking 'missed' is is the better choice.
